OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,1142 @@
+#include <SimTKcommon.h>
+
+
+namespace OpenSim {
+  class AbstractDataTable;
+  template<typename ET> class DataTable_;
+
+  using DataTable = DataTable_<SimTK::Real>;
+
+  /// Enum to specify if the the input iterator is traversing data row-wise or
+  /// col-wise.
+  /// clang3.6 crashes if this is turned to a "enum class". Until it is fixed,
+  /// use a pre c++11 enum.
+  enum InputItDir{
+    ROWWISE, 
+    COLWISE
+  };
+
+  /// Bind two DataTables by row.
+  template<typename ET>
+  DataTable_<ET> rbindDataTables(const DataTable_<ET>& dt1, 
+                                 const DataTable_<ET>& dt2);
+
+  /// Bind two DataTables by col.
+  template<typename ET>
+  DataTable_<ET> cbindDataTables(const DataTable_<ET>& dt1, 
+                                 const DataTable_<ET>& dt2);
+
+  // Exceptions.
+  class ColumnDoesNotExist;
+  class ColumnHasLabel;
+  class ColumnHasNoLabel;
+  class ZeroElements;
+  class InvalidEntry;
+  class MetaDataKeyExists;
+  class MetaDataKeyDoesNotExist;
+  class MetaDataTypeMismatch;
+} // namespace OpenSim
+
+
+/**----------------------------------------------------------------------------*
+* AbstractDataTable is the base-class of all DataTable_ allowing storage of    *
+* DataTable_ of different types to be stored in containers. See Datatable_ for *
+* details on the interface.                                                    *
+*-----------------------------------------------------------------------------*/
+class OpenSim::AbstractDataTable {
+public:
+  AbstractDataTable() = default;
+  AbstractDataTable(const AbstractDataTable&) = delete;
+  AbstractDataTable(AbstractDataTable&&) = delete;
+  virtual std::unique_ptr<AbstractDataTable> clone() const = 0;
+  virtual ~AbstractDataTable() {}
+}; // AbstractDataTable
+
+class OpenSim::ColumnDoesNotExist : public std::runtime_error {
+public:
+  ColumnDoesNotExist(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::ColumnHasLabel : public std::runtime_error {
+public:
+  ColumnHasLabel(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::ColumnHasNoLabel : public std::runtime_error {
+public:
+  ColumnHasNoLabel(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::ZeroElements : public std::runtime_error {
+public:
+  ZeroElements(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::InvalidEntry : public std::runtime_error {
+public:
+  InvalidEntry(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::MetaDataKeyExists : public std::runtime_error {
+public:
+  MetaDataKeyExists(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::MetaDataKeyDoesNotExist : public std::runtime_error {
+public:
+  MetaDataKeyDoesNotExist(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::MetaDataTypeMismatch : public std::runtime_error {
+public:
+  MetaDataTypeMismatch(const std::string& expl) : runtime_error(expl) {}
+};
```

Comment:
I couldn’t find this on github so I’m replying in email.

I don’t have a strong opinion about what is done with exceptions. I kinda think it makes sense for us to derive OpenSim::Exception from std::exception, which will allow us to use our exception classes to do some generic things like print stack traces, or to give a message “This is a bug, please file an issue…”

> Users (and the libraries they probably use) will likely to know std::exception hierarchy 
I don’t think this is true of our users. I think our average user knows what an exception is, but not much more. Your argument is still strong without this point though.

> On Jun 24, 2015, at 11:55 PM, Shrinidhi KL <notifications@github.com <mailto:notifications@github.com>> wrote:
> 
> In OpenSim/Common/DataTable.h <https://github.com/opensim-org/opensim-core/pull/499#discussion_r33226842>:
> 
> > +};
> > +
> > +class OpenSim::MetaDataKeyExists : public std::runtime_error {
> > +public:
> > +  MetaDataKeyExists(const std::string& expl) : runtime_error(expl) {}
> > +};
> > +
> > +class OpenSim::MetaDataKeyDoesNotExist : public std::runtime_error {
> > +public:
> > +  MetaDataKeyDoesNotExist(const std::string& expl) : runtime_error(expl) {}
> > +};
> > +
> > +class OpenSim::MetaDataTypeMismatch : public std::runtime_error {
> > +public:
> > +  MetaDataTypeMismatch(const std::string& expl) : runtime_error(expl) {}
> > +};
> I am looking at the exceptions again and I am now not seeing any benefit to deriving from OpenSim::Exception besides being consistent with rest of the code in OpenSim. I would like to change it to use standard library more because:
> 
> OpenSim::Exception is not doing anything special and I am not sure what will be added to it in future to make it special.
> Users (and the libraries they probably use) will likely to know std::exception hierarchy so deriving our exceptions from std::runtime_error, std::logic_error and so on makes the exceptions more meaningful to me.
> If we are going to transition to using standard library more, this can be start of the transition.
> —
> Reply to this email directly or view it on GitHub <https://github.com/opensim-org/opensim-core/pull/499/files#r33226842>.
>